src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,97 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+    private static final String NO_SUCH_ELEMENT_MESSAGE = "Element not found: ";
+    private static final String INDEX_OUT_OF_BOUND_MESSAGE = "Index out of bound: ";
+    private Object[] elementData = new Object[DEFAULT_CAPACITY];
+    private int size;
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            growIfNeeded();
+        }
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        growIfNeeded();
+        if (index != size) {
+            checkIndex(index);
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        }
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (size + list.size() >= elementData.length) {
+            growIfNeeded();
+        }
```

Comment:
there is redundant call. If you have `growIfNeeded` with capacity is a parameter, It will not be redundant.